PR: Discount field in line items and order summary section

mechaadi opened the pull request:
This will enable us to pass `discount` parameter in the `items` array as well `orderDiscount` parameter in the `invoice` object.

I've also generated and attached the new `example/invoice.pdf` in this PR to show how it looks with these two parameters.

mechaadi committed:
feat: added discount field in line items and summary

mechaadi committed:
feat: order discount argument

mechaadi committed:
feat: added new params in readme

h1dd3nsn1p3r commented:
Hi @mechaadi 

Amazing work :)

If you can look at these suggestions, it would be great.

- The total discount row is always rendered, even when orderDiscount is undefined or zero. Consider conditionally adding this row only when a non-zero discount is provided.
- The `calcTotalDiscount` method currently just returns the input discount. Consider inlining this call or extending it to handle compound discounts to avoid an extra indirection.

mechaadi committed:
feat: visibility of discount in summary if discount is a positive number and removed calcTotalDiscount func

mechaadi commented:
Hello @h1dd3nsn1p3r, great library!

I've taken both review comments into consideration and pushed the changes, let me know how it looks!

mechaadi committed:
chore: updated order discount to show it in the invoice

h1dd3nsn1p3r committed:
:package: Added: [Config string option] Grand total.

- :package: Added: [Config string option] Grand total.
- :package: Tweak: Grand total is now bold and has light gray background.

Why?

We have so many "total", in items table, sub-total, total discount, total tax, etc. "Grand total" is the final amount.

h1dd3nsn1p3r commented:
I renamed the final total to grand total. If grandTotal is not supplied via config a fallback string "Total" will be displayed. 

![image](https://github.com/user-attachments/assets/d8a49ba0-4f9a-4bff-8d86-bf72ba41dcc7)

We are good to merge the PR but I think we can discuss this before doing that.

🤔 I am not sure if I now like how the `tax %`, `discount amount` and `total amount` are displayed in the items table column. I am not talking about UI, but the logic of how the total amount is being calculated simply by multiplying the quantity and amount. Tax percentage and discount amount appear there just for information. 

The `orderDiscount`, Is this even required if we display the total discount amount? orderDiscount can be renamed/useful in the following use cases: 

- Allow option to allow flat discount on sub-total amount.
- Allow option to allow additional discount after per-item discount.  

What do you think about this?

mechaadi commented:
I agree, for the `total` column in the table, maybe we can use `tax` and `discount` to calculate the *real* total price of the line item.

`orderDiscount` is there because there are two type of discounts:
1. product discount
2. order discount

An example of it is how [shopify](https://help.shopify.com/en/manual/discounts/discount-combinations) has 3 different kind of discounts.

h1dd3nsn1p3r commented:
Great!

So, how about if we do something like this: 

### Items total amount (per item)

Let's convert the discount to a percentage from the amount so the calculation will look like this: 

```javascript
/**
 * Applies a discount (if applicable) to an item based on its quantity and price.
 *
 * Then adds tax to the discounted subtotal if applicable.
 *
 * @returns {number} The total amount for the item after applying discount and tax.
 */
function calculateItemTotal(item) {
  const subtotal = item.quantity * item.price;
  const discountAmount = subtotal * (item.discount / 100);
  const discountedSubtotal = subtotal - discountAmount;
  const taxAmount = discountedSubtotal * (item.tax / 100);
  return discountedSubtotal + taxAmount;
}
```

### Invoice Sub-Total and Grand Total section

In this section, let's display the following: 

- Subtotal from all items total amount.
- Total value of the tax amount (Just information) - _Your input on this, what do you think?_
- Additional or Order discount (if applicable) - This can be supplied via the invoice object. (currently implemented)

```javascript
invoice: {
	...
	orderDiscount: 2, // 2% discount on the total sub-total amount. (Has to be a percentage)
	...
},
```

The Additional or Order discount percentage calculated value should be displayed in the international currency format. 
- Finally, display the Grand Total (after additional/order discount, deducting from the sub-total amount) 

Let me know your thoughts.

mechaadi commented:
Hey there, good suggestions! I am hoping to make some time next week and handle it.

h1dd3nsn1p3r commented:
Great! I'll review the PR once you're done.

h1dd3nsn1p3r commented:
Hi @mechaadi

Any update on this?

sonnguyencimigo commented:
Waiting on this

mechaadi commented:
> Hi @mechaadi
> 
> 
> 
> Any update on this?
> 
> 

Hey, i am getting very limited time these days. Sorry for the continuous delay, i will try my best to push it this week hopefully.

mechaadi committed:
feat: added fees argument

mechaadi committed:
fix: item total calulation by taking tax and discount into the calculation

mechaadi committed:
feat: converted item discount value to percentage

mechaadi commented:
@h1dd3nsn1p3r hey, done with all the suggestions.

As a bonus, I've also added another optional cell in the invoice summary called `fees`. Sometimes payment processors add a fee to the transaction which can be handled with this optional parameter.

Let me know what you think!

mechaadi commented:
Hey @h1dd3nsn1p3r 

Just wanted to check how the recent changes are looking.

We good to merge?

h1dd3nsn1p3r commented:
Hey @mechaadi 

Sorry for the delay. Somehow, I didn't notice additional commits in the PR.  I'll check the changes within this week.

Thank you for the changes.

h1dd3nsn1p3r committed:
:bug: Fix: TAX being applied in the final total amount.

- :bug: Fix: TAX being applied in the final total amount.

h1dd3nsn1p3r committed:
:package: Updated: README.

- :package: Updated: README.

h1dd3nsn1p3r commented:
@mechaadi thank you for your amazing work! Release created: https://github.com/h1dd3nsn1p3r/pdf-invoice/releases

mechaadi commented:
@h1dd3nsn1p3r Amazing, great news! Thanks for working alongside me with this new feature, this library is great.

Do you think we should update the invoice image in the README.md file to highlight the item discount column and discount in order summary section?

h1dd3nsn1p3r commented:
Hi @mechaadi Yes, I’ll update the image shortly.
